### ISSUE TITLE
False positives triggered by missing type information

### ISSUE THREAD

**Issue by zivanfi** [microsoft/pyright#558]
Updated bug report (see the original one at the bottom):

**Describe the bug**

If type information is missing for class high in the inheritance chain, subclasses are considered to be of Any type, triggering false positives when code depends on these subclasses being different from other types.

**Reproduction**

`pip3 install overrides`, then use the following code:

```
from overrides import EnforceOverrides


class Attribute(EnforceOverrides):
    pass


class Color(Attribute):
    pass


class Shape(Attribute):
    pass


value: Attribute = Shape()
if isinstance(value, Color):
    pass
elif isinstance(value, Shape):
    shape: Shape = value
```

Pyright gives the following output:

```
  17:4 - warning: Unnecessary isinstance call: 'Attribute' is always instance of 'Color' (reportUnnecessaryIsInstance)
  20:20 - error: Expression of type 'Never' cannot be assigned to declared type 'Shape'
  'Never' cannot be assigned to 'Shape'
```
The error is shown without any additional configuration, the warning only shows up with `"reportUnnecessaryIsInstance": "warning"` in `pyrightconfig.json`.

`EnforceOverrides` comes from the `overrides` package (version 2.8.0) and has no type information.

**Expected result**

Regardless of whether type information is available for a high-level class of the inheritence chain or not, I would expect pyright to respect the bottom part of the class hierarchy that I defined, according to which Attribute, Shape and Color are different classes.

**VS Code extension or command-line**

Pyright 1.1.26, both VS Code extension or command-line usage.

<details><summary>Original bug report (click to show)</summary>

```
from abc import ABC, abstractmethod

from overrides import EnforceOverrides, overrides


class Attribute(EnforceOverrides):
    pass


class Color(Attribute):
    pass


class Shape(Attribute, ABC):
    @abstractmethod
    def get_area(self) -> float:
        pass


class Square(Shape):
    side_length: float

    def __init__(self, side_length: float):
        self.side_length = side_length

    @overrides
    def get_area(self) -> float:
        return self.side_length ** 2


value: Attribute = Square(1.5)
if isinstance(value, Color):
    print("I'm a Color!")
elif isinstance(value, Shape):
    shape: Shape = value
    area = shape.get_area()
    print(f"I'm a Shape with area {area}!")
```

Pyright gives the following output:

```
  32:4 - warning: Unnecessary isinstance call: 'Attribute' is always instance of 'Color' (reportUnnecessaryIsInstance)
  35:20 - error: Expression of type 'Never' cannot be assigned to declared type 'Shape'
  'Never' cannot be assigned to 'Shape'
```

But `value` is clearly a `Shape`, as shown by the output of executing the program:

```
I'm a Shape with area 2.25!
```

The error goes away if I don't subclass `Attribute` from `EnforceOverrides`, so there may be something in there that confuses pyright into thinking that an abstract class can not be subclassed. Regardless of what `EnforceOverrides` exactly contains though, the rest of the class hierarchy is valid and should be accepted. `EnforceOverrides` comes from the `overrides` package (version 2.8.0).

**To Reproduce**

`pip3 install overrides` then use the example code provided above. The error is shown without any additional configuration, the warning only shows up with `"reportUnnecessaryIsInstance": "warning"` in `pyrightconfig.json`.
</details>

**Comment by erictraut** [comment]
Thanks for the bug report.

The problem stems from the fact that you don't have a type stub installed for "overrides" and you don't have the "pyright.useLibraryImplementations" setting enabled, so there's no type information available for the EnforceOverrides class. In that case, Pyright assigns it an "Unknown" type (which is effectively the same as "Any"). This is causing a problem for the "unnecessary isinstance" logic.

If the type is Unknown or Any, I should probably disable type narrowing and the "unnecessary isinstance" check.

**Comment by zivanfi** [comment]
> there's no type information available for the EnforceOverrides class

Yes, but please note that `Attribute`, `Color`, `Shape` and `Square` are are all subclasses of `EnforceOverrides`, yet pyright thinks that an `Attribute` can only be `Color`. Regardless of whether type information is available or not, I would expect pyright to respect the class hierarchy that I defined, according to which `Square` can be an `Attribute` too.

**Comment by zivanfi** [comment]
Ah, okay, I think I finally understand what you meant. Essentially, pyright considers the first condition to be equivalent to `if isinstance(value, Any)`, regardless of the type. So there is no asymmetry regarding the involved classes, what I perceived to be an asymmetry came from the order of the type checks: If I check for `isinstance(value, Shape)` first, then that will be considered to be an always true unnecessary check. Similarly, if I put the second condition in a separate `if` instead of an `elif`, then it won't be considered an impossible case and it will get rid of the error (but not the warning).

Also, the error has nothing to do with the abstractness of one the classes as I mistakenly believed and repo example can be significantly simplified.

**Comment by zivanfi** [comment]
I updated the bug description based on your explanation.

**Comment by erictraut** [comment]
Yes, your understanding is correct. When no type information is available, Pyright has no idea whether the symbol represents a class, an object, a function, a type variable, a type alias, etc. You could argue that it should be able to assume that "EnforceOverrides" is a class of some sort given how it's used within this source file, but inferring types based on the usage of a symbol is challenging.

I'll work on eliminating the false positives you're seeing.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.27, which I just published.
